stomp_moveit/src/stomp_smoothing_adapter.cpp

```diff
@@ -0,0 +1,139 @@
+                    const planning_interface::MotionPlanRequest& req, planning_interface::MotionPlanResponse& res,
+                    std::vector<std::size_t>& added_path_index) const override
+  {
+    // Following call to planner() calls the OMPL planner and stores the trajectory inside the MotionPlanResponse res
+    // variable which is then used by STOMP for optimization of the computed trajectory
+    if (!planner(ps, req, res))
+      return false;
+
+    // STOMP reads the seed trajectory from trajectory constraints so we need to convert the waypoints first
+    const size_t seed_waypoint_count = res.trajectory_->getWayPointCount();
+    const std::vector<std::string> joint_names =
+      res.trajectory_->getFirstWayPoint().getJointModelGroup(req.group_name)->getActiveJointModelNames();
+    const size_t joint_count = joint_names.size();
+    planning_interface::MotionPlanRequest seed_req = req;
+    seed_req.trajectory_constraints.constraints.clear();
+    seed_req.trajectory_constraints.constraints.resize(seed_waypoint_count);
+    for (size_t i = 0; i < seed_waypoint_count; ++i)
+    {
+      seed_req.trajectory_constraints.constraints[i].joint_constraints.resize(joint_count);
+      for (size_t j = 0; j < joint_count; ++j)
+      {
+        seed_req.trajectory_constraints.constraints[i].joint_constraints[j].joint_name = joint_names[j];
+        seed_req.trajectory_constraints.constraints[i].joint_constraints[j].position =
+          res.trajectory_->getWayPoint(i).getVariablePosition(joint_names[j]);
+      }
+    }
+
+    // Get group config
+    const auto& group_config_it = group_config_.find(req.group_name);
+    if (group_config_it == group_config_.end())
+    {
+      ROS_ERROR_STREAM("STOMP is not configured for planning group " << req.group_name);
+      res.error_code_.val = moveit_msgs::MoveItErrorCodes::FAILURE;
+      return false;
+    }
+
+    // Initialize STOMP Planner
+    stomp_moveit::StompPlanner stompPlanner(req.group_name, group_config_it->second, ps->getRobotModel());
+    if(!stompPlanner.canServiceRequest(seed_req))
+    {
+      ROS_ERROR("STOMP planner unable to service request");
+      res.error_code_.val = moveit_msgs::MoveItErrorCodes::FAILURE;
+      return false;
+    }
+
+    // Setup Planning Context
+    stompPlanner.clear();
+    stompPlanner.setPlanningScene(ps);
+    stompPlanner.setMotionPlanRequest(seed_req);
+
+    // Solve
+    ROS_DEBUG("Smoothing result trajectory with STOMP");
+    planning_interface::MotionPlanDetailedResponse stomp_res;
+    bool success = stompPlanner.solve(stomp_res);
+    if (success)
+    {
+      // Successful responses always contain one entry for trajectory_ and proccessing_time_
+      res.trajectory_ = stomp_res.trajectory_.back();
+      res.planning_time_ += stomp_res.processing_time_.back();
+    }
+    res.error_code_ = stomp_res.error_code_;
+    return success;
```

Comment:
Is a failure in STOMP as a planning adapter really a failure of the entire planning pipeline? If a valid plan is acquired from the nominal planner (like OMPL), and STOMP fails to improve it, it seems like the planning adapter should report some sort of message but still return the successful input plan. I have been using this plugin a little bit over the last couple days, and in many cases my OMPL seed plan succeeds, but STOMP fails to improve it. In my use-case I would prefer to keep the OMPL plan when STOMP fails, but the planning service returns failure. Would this be a beneficial change to make to this planning adapter, or does that violate the general concept behind MoveIt planning adapters?

Reply:
This seems like reasonable behavior, but perhaps this could be done in a separate PR. There is related issue of STOMP (or really any trajectory optimization method that turns hard constraints into costs) that valid initial paths can be turned into cost-optimal but infeasible paths. So in summary:
1. If the cost of the optimized path is no better than the cost of the initial path, or
2. If the optimized path is no longer collision-free, or
3. if STOMP fails/times out for some reason,
then the original path should be returned.

Reply:
Good points. I don't mind creating a separate PR with these changes after this one gets merged. I just wanted to bring up those concerns in case they were relevant to other use-cases or affected this PR in a meaningful way

Reply:
Can we at the very least make this behaviour configurable?

Failing early is a very good thing to do. If users include a smoothing adapter in the motion planning pipeline, they do that for a reason. If that smoothing step is unable to produce a proper result, that should result in a hard fail.

MoveIt used to silently fail and continue stumbling on with whatever it could salvage for a good nr of years of its existence until that behaviour got fixed. Precisely that behaviour is what made it so "unreliable" in the eyes of many (industrial?) users.

Let's not add that again.

Or at least allow users to configure whether they want a hard fail or continue even though some step failed.

Reply:
As stated in my PR above, I don't believe STOMP can fail by design if provided with a valid seed. If it still does that is not a feature that should be configurable but a bug. OMPL can be a bit tolerant in what it considers a valid trajectory (mostly due to discrete collision checking), if STOMP discovers that, it should fail!
I'll try to check if there is a better way of seeding stomp or if there is some way to tell it to use the seed as 0th iteration as once some iteration is collision free, it will discard colliding samples in subsequent iterations.

Could you provide a sample @marip8

Reply:
I think making the behavior configurable (with the default being hard fail early) would add a lot of flexibility and shouldn't be too hard to implement. As mentioned above, this feature doesn't necessarily need to be a part of this PR. I'm happy to add it later in the interest of getting this PR merged.

> As stated in my PR above, I don't believe STOMP can fail by design if provided with a valid seed.

I was initially under this impression as well, but I've been able to trigger STOMP failures with a number of seemingly valid OMPL seed trajectories. One aspect of these failures is that STOMP was not using the OMPL plan as a seed due a potential [IK-solving bug](https://github.com/ros-industrial/stomp_ros/issues/19#issuecomment-622089046) in the planner. After I fixed that locally, the failures I continued to see were basically timeouts because STOMP kept pushing the trajectory further and further away from the valid seed path. I increased the timeout up to 30 seconds at one point, and it still failed to finish within that time

Here is a short screen capture I took that shows this fairly well. The green path is a marker representing the seed state, and the blue trajectory is the STOMP update. In this case I provided the goal as a Cartesian tool pose (through a slight modification to the MoveIt Rviz widget). This seems to occur both when the goal constraints are joint poses or Cartesian tool poses, though. My branch with these testing changes is [here](https://github.com/marip8/stomp_ros/tree/sandbox/planning_adapter).

![test](https://user-images.githubusercontent.com/18411310/80813878-7f914600-8b90-11ea-8cd8-52689f4dcbba.gif)

<details>
<summary>STOMP Parameters</summary>
<p><pre><code>
stomp/manipulator:
  group_name: lift_manipulator
  optimization:
    num_timesteps: 25
    num_iterations: 100
    num_iterations_after_valid: 0
    num_rollouts: 50
    max_rollouts: 50
    initialization_method: 1 #[1 : LINEAR_INTERPOLATION, 2 : CUBIC_POLYNOMIAL, 3 : MININUM_CONTROL_COST]
    control_cost_weight: 0.0
  task:
    noise_generator:
      - class: stomp_moveit/GoalGuidedMultivariateGaussian
        stddev: [0.25, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]
    cost_functions:
      - class: stomp_moveit/CollisionCheck
        collision_penalty: 1.0
        cost_weight: 1.0
        kernel_window_percentage: 0.4
        longest_valid_joint_move: 0.025
      - class: stomp_moveit/ToolGoalPose
        position_cost_weight: 0.5
        orientation_cost_weight: 0.5
    noisy_filters:
      - class: stomp_moveit/JointLimits
        lock_start: True
        lock_goal: False
      - class: stomp_moveit/MultiTrajectoryVisualization
        line_width: 0.02
        rgb: [255, 255, 0]
        marker_array_topic: stomp_trajectories
        marker_namespace: noisy
    update_filters:
      - class: stomp_moveit/ConstrainedCartesianGoal
      - class: stomp_moveit/PolynomialSmoother
        poly_order: 5
      - class: stomp_moveit/TrajectoryVisualization
        line_width: 0.02
        rgb: [0, 0, 255]
        error_rgb: [255, 0, 0]
        publish_intermediate: True
        marker_topic: stomp_trajectory
        marker_namespace: optimized
    visualization:
      - class: stomp_moveit/TrajectoryVisualization
        line_width: 0.02
        rgb: [0, 255, 0]
        error_rgb: [255, 0, 0]
        publish_intermediate: True
        marker_topic: seed
        marker_namespace: seed
</code></pre></p>
</details>

Maybe it's just a parameter tuning problem, but this is what sparked my initial comment above

Reply:
@simonschmeisser, you say that you don't believe that STOMP cannot fail by design, but finding local minima of a cost function in trajectory space and finding a collision-free path are two different problems. A STOMP-produced trajectory can be both optimal and infeasible (at least in theory) if going through a thin obstacle has lower cost than going a long way around it. With careful shaping of the cost function this kind of issue can be mitigated but not eliminated.

Anecdotally, I have seen that STOMP often takes a long time to converge, even starting from a reasonably smooth & feasible initial trajectory.

Reply:
There is some code commented out that allows you to visualize the rollouts. I think it fails because the sampling region for your goal is to big (that is another open bug). If you enable more debug output and the visualization you should see a number of spheres around the goal that resemble the sampled goal. They typically look wrong (badly distributed, too far from goal). You might want to try `lock_goal: True`

Also if you had valid goals you'd probably have valid solutions as well and this would take only one iteration. I usually put something larger than 0 for `iterations_after_valid`, eg 5

Finally visualizing the rollouts should allow you to tune the noise_generator

Reply:
@mamoll My statement only holds under the precondition of a valid (collision free) seed. It then discards iterations that change from valid to invalid. So worst case you should get the seed as a result.

This is of course a slightly theoretical statement as there are bugs ... (especially with the goal sampling and handling)

Reply:
I agree with @simonschmeisser, if during an optimization step Stomp can't find a valid solution then it should drop it, in the case of starting with a valid collision free seed it should then return that in the worst case so the fact that is not is indicative of a bug or perhaps a bad configuration, judging from @marip8 configuration nothing really stands out to me.

@marip8 @gavanderhoorn making the hardfail configurable sounds reasonable to me